### PR TITLE
Support new Flax linen modules

### PR DESCRIPTION
**Pull request by gustavehug**
Closes https://github.com/pyro-ppl/numpyro/issues/979

Note that users now have to prefix e.g. priors with `params` with linen.

### CODE REVIEW

File: numpyro/contrib/module.py [hunk 1]
```diff
@@ -49,12 +49,16 @@ def flax_module(name, nn_module, *, input_shape=None, **kwargs):
...
         # feed in dummy data to init params
         rng_key = numpyro.prng_key()
-        _, nn_params = nn_module.init(rng_key, *args, **kwargs)
+        nn_params = nn_module.init(rng_key, *args, **kwargs)
+
+        # make a mutable copy
+        nn_params = flax.core.unfreeze(nn_params)
+
         # make sure that nn_params keep the same order after unflatten
         params_flat, tree_def = tree_flatten(nn_params)
```
Comment (fehiepsi) [inline numpyro/contrib/module.py:58]: I guess we only need to extract `params` field for optimizer:
```python
    params_flat, tree_def = tree_flatten(nn_params["params"])
    ...
return partial(nn_module.apply, {"params": nn_params})
```
This will keep the params ingestion the same as before (i.e. no prefix `params`). However, this will throw away important mutable variables like `batch_stats` in [nn.BatchNorm](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html#Variables-and-collections-of-variables).

I'm not sure what's a good solution here. Probably it is better to stick with the old pattern (of setting priors) and think about how to address extra mutable stats in another PR. What do you think?

Comment (fehiepsi) [reply numpyro/contrib/module.py:58]: One solution to address extra mutable variables is to define inside this `flax_module` primitive something like this
```python
def apply_fn(*args, **kwargs_)
    variables = {"params": nn_params}
    mutable_vars = kwargs.get("mutable", [])
    if mutable_vars:
        for var in mutable_vars:
            # get the current mutable value
            mutable = numpyro.param_like(name + "$" + var)
            variables[var] = mutable  # jax.lax.stop_gradient(mutable)
        out, variables = nn_module.apply(variables, *args, **kwargs_)
        for var, mutable in variables.items():
            # update the current mutable value
            numpyro.param_like(name + "$" + var, mutable)
    else:
        out = nn_module.apply(variables, *args, **kwargs_)
    return out

return apply_fn
```
Here `SVI` will recognize those variables defined in `param_like` primitives and store it in `SVIState`.

**Comment by fehiepsi** [comment]
Hi @gustavehug, I'm going to make some enhancements to address some issues with BatchNorm and Dropout layers. Would you mind resolving the new pattern for param names in this PR?

File: numpyro/contrib/module.py [hunk 2]
```diff
@@ -49,12 +49,16 @@ def flax_module(name, nn_module, *, input_shape=None, **kwargs):
         args = (jnp.ones(input_shape),) if input_shape is not None else ()
         # feed in dummy data to init params
         rng_key = numpyro.prng_key()
-        _, nn_params = nn_module.init(rng_key, *args, **kwargs)
+        nn_params = nn_module.init(rng_key, *args, **kwargs)
+
+        # make a mutable copy
+        nn_params = flax.core.unfreeze(nn_params)
```
Comment (fehiepsi) [inline numpyro/contrib/module.py:55]: You can capture parameters with `nn_params = flax.core.unfreeze(nn_params)["params"]`

File: numpyro/contrib/module.py [hunk 3]
```diff
@@ -49,12 +49,16 @@ def flax_module(name, nn_module, *, input_shape=None, **kwargs):
...
+
         # make sure that nn_params keep the same order after unflatten
         params_flat, tree_def = tree_flatten(nn_params)
         nn_params = tree_unflatten(tree_def, params_flat)
         numpyro.param(module_key, nn_params)
-    return partial(nn_module.call, nn_params)
+    return partial(nn_module.apply, nn_params)
```
Comment (fehiepsi) [inline numpyro/contrib/module.py:61]: With `params` captured as above, we can recover it here:
```
return partial(nn_module.apply, {"params": nn_params})
```

File: numpyro/contrib/module.py [hunk 4]
```diff
@@ -142,25 +146,27 @@ def random_flax_module(name, nn_module, prior, *, input_shape=None, **kwargs):
...
 
             class B(nn.Module):
-                def apply(self, x):
-                    return A(x, name='inner')
+                @nn.compact
+                def __call__(self, x):
+                    return A(name='inner')(x)
 
-        has parameters `{'inner': {'dense': {'kernel': param_value}}}`. In the argument
+        has parameters `{'params': {'inner': {'dense': {'kernel': param_value}}}}`. In the argument
```
Comment (fehiepsi) [inline numpyro/contrib/module.py:158]: no need to change the pattern here

File: numpyro/contrib/module.py [hunk 5]
```diff
@@ -142,25 +146,27 @@ def random_flax_module(name, nn_module, prior, *, input_shape=None, **kwargs):
...
         `prior`, to specify `kernel` parameter, we join the path to it using dots:
-        `prior={"inner.dense.kernel": param_prior}`.
+        `prior={"params.inner.dense.kernel": param_prior}`.
```
Comment (fehiepsi) [inline numpyro/contrib/module.py:160]: you can revert the change

File: numpyro/contrib/module.py [hunk 6]
```diff
@@ -142,25 +146,27 @@ def random_flax_module(name, nn_module, prior, *, input_shape=None, **kwargs):
...
-    :param flax.nn.Module: the module to be registered with NumPyro
+    :param flax.linen.Module: the module to be registered with NumPyro
     :param prior: a NumPyro distribution or a Python dict with parameter names as keys and
         respective distributions as values. For example::
 
             net = random_flax_module("net",
-                                     flax.nn.Dense.partial(features=1),
-                                     prior={"bias": dist.Cauchy(), "kernel": dist.Normal()},
+                                     flax.linen.Dense(features=1),
+                                     prior={"params.bias": dist.Cauchy(), "params.kernel": dist.Normal()},
```
Comment (fehiepsi) [inline numpyro/contrib/module.py:169]: you can revert the change here

File: test/contrib/test_module.py [hunk 7]
```diff
@@ -85,13 +85,15 @@ def test_flax_module():
 
     with handlers.trace() as flax_tr, handlers.seed(rng_seed=1):
         flax_model_by_shape(X, Y)
-    assert flax_tr["nn$params"]["value"]["kernel"].shape == (100, 100)
-    assert flax_tr["nn$params"]["value"]["bias"].shape == (100,)
+    params = flax_tr["nn$params"]["value"]["params"]
```
Comment (fehiepsi) [inline test/contrib/test_module.py:88]: you can either revert the change or removing `["params"]` here

File: test/contrib/test_module.py [hunk 8]
```diff
@@ -85,13 +85,15 @@ def test_flax_module():
...
+    assert params["kernel"].shape == (100, 100)
+    assert params["bias"].shape == (100,)
 
     with handlers.trace() as flax_tr, handlers.seed(rng_seed=1):
         flax_model_by_kwargs(X, Y)
-    assert flax_tr["nn$params"]["value"]["kernel"].shape == (100, 100)
-    assert flax_tr["nn$params"]["value"]["bias"].shape == (100,)
+    params = flax_tr["nn$params"]["value"]["params"]
```
Comment (fehiepsi) [inline test/contrib/test_module.py:94]: same as above

File: test/contrib/test_module.py [hunk 9]
```diff
@@ -154,9 +156,9 @@ def test_random_module__mcmc(backend, init):
     if backend == "flax":
         import flax
 
-        linear_module = flax.nn.Dense.partial(features=1)
-        bias_name = "bias"
-        weight_name = "kernel"
+        linear_module = flax.linen.Dense(features=1)
+        bias_name = "params.bias"
+        weight_name = "params.kernel"
```
Comment (fehiepsi) [inline test/contrib/test_module.py:161]: you can use the old param names here

**Comment by fehiepsi** [comment]
@gustavehug In #1016, I just add supports for mutable states. When this is merged, I'll add an LDA example that utilizes `flax.linen`. :)

**Comment by fehiepsi** [comment]
Hi @gustavehug, can I push some commits to this PR to address the above API issues?

**Comment by fehiepsi** [comment]
Thank you, @gustavehug, for the update. We have incorporated part of your changes in the master branch. :)
